### ISSUE TITLE
Sequence ID assignment change from 1.5 to 1.6 breaking fprop and bprop correlation

### ISSUE THREAD

**Issue by debermudez** [pytorch/pytorch#42564]
## 🐛 Bug

Pytorch 1.6 is no longer emitting sequence ids that allow for correlation between forward and backward propagation using the torch.autograd.profiler.emit_nvtx context manager. 

We are seeing the following behavior:

1. Operations are associated with multiple sequence ids. 
2. Sequence Ids are being reused from one operation to another.

## To Reproduce

Using the following code snippet, you can reproduce the error we are seeing.

```#!/usr/bin/env python3
import torch
import torch.nn as nn
import torch.cuda.profiler as profiler
N = 32
I = 128 
H = 256 
O = 1024
# 2 Layer MLP
model = torch.nn.Sequential(
            torch.nn.Linear(I, H), 
            torch.nn.Linear(H, O)
        ).cuda().half()
# Input and Label
x = torch.randn(N, I).cuda().half()
target = torch.empty(N, dtype=torch.long).random_(O).cuda()
# Loss and optimizer
criterion = nn.CrossEntropyLoss().cuda()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
optimizer.zero_grad()
with torch.autograd.profiler.emit_nvtx():
    profiler.start()
    output = model(x)
    loss = criterion(output, target)
    loss.backward()
    optimizer.step()
    profiler.stop()
```
To profile:
```
nvprof -fo a.nvvp --profile-from-start off python a.py
```

## Expected behavior

With PyTorch 1.5, the seq ids are easy to understand and associate.
```
Linear 1: legacy::elementwise_kernel : seqId = 1
Linear 1: turing_fp16_s884gemm_fp16_64x64_ldg8_f2f_tn : seqId = 1
Linear 2: legacy::elementwise_kernel: seqId = 3
Linear 2: turing_s884gemm_fp16_64x64_ldg8_tn : seqId = 3
Linear 2: splitKreduce_kernel : seqId = 3
Softmax : softmax_warp_forward : seqId = 4
and so on.
```
## Current Pytorch 1.6 behavior:
With PyTorch 1.6, the seq id assignment has changed and there is some sort of overlapping.
```
Linear 1: modern::unrolled_elementwise_kernel : seqId = 1,2
Linear 1: cutlass::Kernel : seqId = 1,2
Linear 2: modern::unrolled_elementwise_kernel: seqId = 3,4
Linear 2: cutlass::Kernel : seqId = 3,4
Softmax : softmax_warp_forward : seqId = 4,5
and so on.
```

For back prop,  it is not possible to tell if seq id = 4 is associated with linear or softmax. This is just one of many examples we now run into when executing networks on Pytorch 1.6.

## Environment
The Pytorch 1.5 version:
```PyTorch version: 1.5.0a0+8f84ded
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: TITAN RTX
Nvidia driver version: 450.27
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] msgpack-numpy==0.4.3.2
[pip] numpy==1.18.1
[pip] pytorch-transformers==1.1.0
[pip] torch==1.5.0a0+8f84ded
[pip] torchtext==0.4.0
[pip] torchvision==0.6.0a0
[conda] magma-cuda101             2.5.1                         1    local
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] msgpack-numpy             0.4.3.2                  py36_0  
[conda] nomkl                     3.0                           0  
[conda] numpy                     1.18.1           py36h94c655d_0  
[conda] numpy-base                1.18.1           py36h2f8d375_1  
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] torch                     1.5.0a0+8f84ded          pypi_0    pypi
[conda] torchtext                 0.4.0                    pypi_0    pypi
[conda] torchvision               0.6.0a0                  pypi_0    pypi
```

The Pytorch 1.6 version:
```
PyTorch version: 1.6.0a0+9907a3e
Is debug build: No
CUDA used to build PyTorch: 11.0

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: TITAN RTX
Nvidia driver version: 450.27
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.1

Versions of relevant libraries:
[pip] msgpack-numpy==0.4.3.2
[pip] numpy==1.18.1
[pip] pytorch-transformers==1.1.0
[pip] torch==1.6.0a0+9907a3e
[pip] torchtext==0.6.0
[pip] torchvision==0.7.0a0
[conda] magma-cuda110             2.5.2                         5    local
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] msgpack-numpy             0.4.3.2                  py36_0  
[conda] nomkl                     3.0                           0  
[conda] numpy                     1.18.1           py36h94c655d_0  
[conda] numpy-base                1.18.1           py36h2f8d375_1  
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] torch                     1.6.0a0+9907a3e          pypi_0    pypi
[conda] torchtext                 0.6.0                    pypi_0    pypi
[conda] torchvision               0.7.0a0                  pypi_0    pypi
```

credit to @adityaiitb for originally finding this issue

**Comment by mcarilli** [comment]
According to @ilia-cher on Pytorch slack, fix is PRed here https://github.com/pytorch/pytorch/pull/42565.

**Comment by debermudez** [comment]
Forgot to close this. This was fixed some time ago. Thank you for the hard work.
